File: stack/020_valid_parentheses.py
"""
Problem: Valid Parentheses
LeetCode #: 20
Difficulty: Easy
Link: https://leetcode.com/problems/valid-parentheses/

Approach: Use a stack to match opening/closing brackets.
Time Complexity: O(n)
Space Complexity: O(n)
"""


class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {")": "(", "}": "{", "]": "["}

        for char in s:
            if char in mapping:
                top = stack.pop() if stack else "#"
                if mapping[char] != top:
                    return False
            else:
                stack.append(char)

        return not stack


# --- Test ---
if __name__ == "__main__":
    sol = Solution()
    print(sol.isValid("()"))      # True
    print(sol.isValid("()[]{}"))  # True
    print(sol.isValid("(]"))      # False

File: stack/020_valid_parentheses.java
/*
 * Problem: Valid Parentheses
 * LeetCode #: 20
 * Difficulty: Easy
 * Link: https://leetcode.com/problems/valid-parentheses/
 *
 * Approach: Stack matching closing brackets to opening brackets.
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

import java.util.Stack;

public class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == '(') stack.push(')');
            else if (c == '{') stack.push('}');
            else if (c == '[') stack.push(']');
            else if (stack.isEmpty() || stack.pop() != c) return false;
        }
        return stack.isEmpty();
    }

    // --- Test ---
    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.isValid("()[]{}")); // true
        System.out.println(sol.isValid("(]"));     // false
    }
}
